Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertMaterialQuantities.cs

```diff
@@ -11,69 +10,148 @@ namespace Objects.Converter.Revit
+    {
+      foreach (ElementId matId in element.GetMaterialIds(false))
+      {
+        double volume = element.GetMaterialVolume(matId);
+        double area = element.GetMaterialArea(matId, false);
+        yield return Create(element, matId, area, volume, units);
+      }
+    }
+
+    private MaterialQuantity? GetMaterialQuantityForMEPElement(DB.Element element, string units)
+    {
+      DB.Material material = GetMEPSystemRevitMaterial(element);
+      if (material == null)
+      {
         return null;
+      }
 
-      // To-Do: These methods from the Revit API appear to have bugs.
-      double volume = element.GetMaterialVolume(material.Id);
-      double area = element.GetMaterialArea(material.Id, false);
+      DB.Options options = new() { DetailLevel = ViewDetailLevel.Fine };
+      var (solids, _) = GetSolidsAndMeshesFromElement(element, options);
 
-      // Convert revit interal units to speckle commit units
-      double factor = ScaleToSpeckle(1);
-      volume *= factor * factor * factor;
-      area *= factor * factor;
+      var (area, volume) = GetAreaAndVolumeFromSolids(solids);
+      return Create(element, material.Id, area, volume, units);
+    }
+    
+    private IEnumerable<MaterialQuantity> GetMaterialQuantitiesFromSolids(DB.Element element, string units)
+    {
+      DB.Options options = new() { DetailLevel = ViewDetailLevel.Fine };
+      var (solids, _) = GetSolidsAndMeshesFromElement(element, options);
+
+      foreach (ElementId matId in GetMaterialsFromSolids(solids))
+      {
+        var (area, volume) = GetAreaAndVolumeFromSolids(solids, matId);
+        yield return Create(element, matId, area, volume, units);
+      }
+    }
 
-      var speckleMaterial = ConvertAndCacheMaterial(material.Id, material.Document);
-      var materialQuantity = new Objects.Other.MaterialQuantity(speckleMaterial, volume, area, units);
+    private MaterialQuantity Create(
+      Element element,
+      ElementId materialId,
+      double areaRevitInternalUnits,
+      double volumeRevitInternalUnits,
+      string units)
+    {
+      Other.Material speckleMaterial = ConvertAndCacheMaterial(materialId, element.Document);
+      double factor = ScaleToSpeckle(1);
+      double area = factor * factor * areaRevitInternalUnits;
+      double volume = factor * factor * factor * volumeRevitInternalUnits;
+      MaterialQuantity materialQuantity = new(speckleMaterial, volume, area, units);
 
       if (LocationToSpeckle(element) is ICurve curve)
+      {
         materialQuantity["length"] = curve.length;
+      }
+      else if (element is DB.Architecture.Railing)
+      {
+        materialQuantity["length"] = (element as DB.Architecture.Railing).GetPath().Sum(e => e.Length) * factor;
+      }
+      else if (element is DB.Architecture.ContinuousRail)
+      {
+        materialQuantity["length"] = (element as DB.Architecture.ContinuousRail).GetPath().Sum(e => e.Length) * factor;
+      }
+
       return materialQuantity;
```

Comment:
what if there's no match? I wonder if a switch on element would be better with default containg the if (LocationToSpeckle(element) is ICurve curve) and if it doesn't match....  should it just return the default MaterialQuantity or should it explode?

Reply:
Changed to switch. Definitely no need for the conversion to explode if one of these criteria is not met